mlflow/store/artifact/optimized_s3_artifact_repo.py

```diff
@@ -0,0 +1,275 @@
+        )
+
+    def _get_write_credential_infos(self, remote_file_paths):
+        """
+        Instead of returning ArtifactCredentialInfo objects, we instead return a list of initialized
+        S3 client. We do so because S3 clients cannot be instantiated within each thread.
+        """
+        return [self._get_s3_client() for _ in remote_file_paths]
+
+    def _upload_to_cloud(self, cloud_credential_info, src_file_path, artifact_file_path):
+        dest_path = posixpath.join(self.bucket_path, artifact_file_path)
+        key = posixpath.normpath(dest_path)
+        if (
+            MLFLOW_ENABLE_MULTIPART_UPLOAD.get()
+            and os.path.getsize(src_file_path) > _MULTIPART_UPLOAD_CHUNK_SIZE
+        ):
+            self._multipart_upload(cloud_credential_info, src_file_path, self.bucket, key)
+        else:
+            self._upload_file(cloud_credential_info, src_file_path, self.bucket, key)
+
+    def _multipart_upload(self, cloud_credential_info, local_file, bucket, key):
+        # Create multipart upload
+        s3_client = cloud_credential_info
+        response = s3_client.create_multipart_upload(Bucket=bucket, Key=key)
+        upload_id = response["UploadId"]
+
+        # Create presigned URL for each part
+        num_parts = math.ceil(os.path.getsize(local_file) / _MULTIPART_UPLOAD_CHUNK_SIZE)
+
+        def _get_presigned_upload_part_url(part_number):
+            return s3_client.generate_presigned_url(
+                "upload_part",
+                Params={
+                    "Bucket": bucket,
+                    "Key": key,
+                    "UploadId": upload_id,
+                    "PartNumber": part_number,
+                },
+            )
+
+        presigned_urls = [
+            _get_presigned_upload_part_url(part_number) for part_number in range(1, num_parts + 1)
+        ]
+
+        # define helper functions for uploading data
+        def _upload_part(presigned_url, local_file, start_byte, size):
+            data = read_chunk(local_file, size, start_byte)
+            with cloud_storage_http_request("put", presigned_url, data=data) as response:
+                augmented_raise_for_status(response)
+                return response.headers["ETag"]
+
+        try:
+            # Upload each part with retries
+            futures = {}
+            for index, presigned_url in enumerate(presigned_urls):
+                part_number = index + 1
+                start_byte = index * _MULTIPART_UPLOAD_CHUNK_SIZE
+                future = self.chunk_thread_pool.submit(
+                    _upload_part,
+                    presigned_url=presigned_url,
+                    local_file=local_file,
+                    start_byte=start_byte,
+                    size=_MULTIPART_UPLOAD_CHUNK_SIZE,
+                )
+                futures[future] = part_number
+
+            results, errors = _complete_futures(futures, local_file)
+            if errors:
+                raise MlflowException(
```

Comment:
Do we want to include the errors like this as part of the exception message?

In `_complete_futures()` we throw away the trace - do we have to?

What if we modified `_complete_futures()` to return the full trace data from the futures completion state?

```python
import traceback

def _complete_futures(futures_dict, file):
    """
    Waits for the completion of all the futures in the given dictionary and returns
    a tuple of two dictionaries. The first dictionary contains the results of the
    futures (unordered) and the second contains the errors (unordered) that occurred
    during the execution of the futures.
    """
    results = {}
    errors = {}

    with ArtifactProgressBar.chunks(
        os.path.getsize(file),
        f"Uploading {file}",
        _MULTIPART_UPLOAD_CHUNK_SIZE,
    ) as pbar:
        for future in as_completed(futures_dict):
            key = futures_dict[future]
            try:
                results[key] = future.result()
                pbar.update()
            except Exception as e:
                # Get the actual underlying trace from each failed future
                traceback_str = traceback.format_exc()
                errors[key] = (repr(e), traceback_str)

    return results, errors
```

And then modify the handler here to persist the trace to aid in debugging if failures occur?

```python
if errors:
    combined_errors = "\n\n---\n\n".join(f"Error: {e}\n\nTraceback:\n{tb}" for e, tb in errors.values())
    message = f"Failed to upload at least one part of {local_file}. Errors encountered:\n{combined_errors}"
    raise MlflowException(message)
```

Reply:
This would be very helpful for debugging for sure. However, I don't think this change needs to be in this PR. It would affect not just the Optimized S3 artifact repository but also the ADLS and Databricks artifact repos. Perhaps you could file a follow PR to make this change?